TombEngine/Objects/TR4/Trap/SquishyBlock.cpp

```diff
@@ -0,0 +1,204 @@
+	{
+		SQUISHY_BLOCK_STATE_MOVE = 0,
+		SQUISHY_BLOCK_STATE_COLLIDE_LEFT = 1,
+		SQUISHY_BLOCK_STATE_COLLIDE_RIGHT = 2,
+		SQUISHY_BLOCK_STATE_BAKED_MOTION = 3,
+		SQUISHY_BLOCK_STATE_COLLIDE_FRONT = 4,
+		SQUISHY_BLOCK_STATE_COLLIDE_BACK = 5
+	};
+
+	enum SquishyBlockAnim
+	{
+		SQUISHY_BLOCK_ANIM_MOVE = 0,
+		SQUISHY_BLOCK_ANIM_IMPACT_BACK = 1,
+		SQUISHY_BLOCK_ANIM_IMPACT_FRONT = 2,
+		SQUISHY_BLOCK_ANIM_BAKED_MOTION = 3,
+		SQUISHY_BLOCK_ANIM_COLLIDE_FRONT = 4,
+		SQUISHY_BLOCK_ANIM_COLLIDE_BACK = 5
+	};
+
+	void InitializeSquishyBlock(short itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		if (!item.TriggerFlags)
+			SetAnimation(item, SQUISHY_BLOCK_ANIM_BAKED_MOTION);
+
+		item.HitPoints = NOT_TARGETABLE;
+		item.ItemFlags[0] = item.TriggerFlags;
+		item.ItemFlags[1] = 0;
+		item.ItemFlags[4] = 0;
+	}
+
+	void ControlSquishyBlock(short itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+		const auto& object = Objects[item.ObjectNumber];
+
+		short& something0 = item.ItemFlags[0];
+		short& someAngle = item.ItemFlags[1];
+		short& headingAngle = item.ItemFlags[4];
+
+		if (!TriggerActive(&item))
+			return;
+
+		if (!item.TriggerFlags)
+		{
+			if (item.Animation.ActiveState != SQUISHY_BLOCK_STATE_BAKED_MOTION)
+				SetAnimation(item, SQUISHY_BLOCK_ANIM_BAKED_MOTION);
+		}
+		else
+		{
+			if (item.Animation.ActiveState == SQUISHY_BLOCK_ANIM_BAKED_MOTION)
+				SetAnimation(item, SQUISHY_BLOCK_STATE_MOVE);
+
+			something0 = item.TriggerFlags;
+
+			if (item.Animation.ActiveState == SQUISHY_BLOCK_STATE_MOVE)
+			{
+				auto forwardDir = EulerAngles(0, item.Pose.Orientation.y + headingAngle, 0).ToDirection();
+
+				auto pointColl = GetCollision(item.Pose.Position, item.RoomNumber, forwardDir, BLOCK(0.5f));
+
+				if (pointColl.RoomNumber != item.RoomNumber)
+					ItemNewRoom(itemNumber, pointColl.RoomNumber);
+
+				if (!IsNextSectorValid(item, forwardDir, BLOCK(0.5f)))
+				{
+					switch (headingAngle)
+					{
+					default:
+					case ANGLE(0.0f):
+						SetAnimation(item, SQUISHY_BLOCK_ANIM_IMPACT_FRONT);
+						break;
+
+					case ANGLE(-180.0f):
+						SetAnimation(item, SQUISHY_BLOCK_ANIM_IMPACT_BACK);
+						break;
+					}
+				}
+				else
+				{
+					float dist = Lerp(item.TriggerFlags / 4, item.TriggerFlags, something0);
+					item.Pose.Position = Geometry::TranslatePoint(item.Pose.Position, forwardDir, dist);
+				}
+			}
+			else
+			{
+				if ((item.Animation.FrameNumber - GetAnimData(item).frameBase) == 19)
+				{
+					if (item.HitPoints != NOT_TARGETABLE && item.HitPoints)
+					{
+						someAngle = item.HitPoints;
+						item.HitPoints = NOT_TARGETABLE;
+					}
+
+					headingAngle += ANGLE(180.0f);
+					item.Pose.Orientation.y += ANGLE(someAngle);
+				}				
+			}
+		}
+
+		AnimateItem(&item);		
+	}
+
+	void ControlFallingSquishyBlock(short itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		short& something0 = item.ItemFlags[0];
+
+		if (TriggerActive(&item))
+		{
+			if (something0 < 60)
```

Comment:
Please try to demagic all the magic numbers here as much as possible.

Reply:
fixed